### PR TITLE
Refine accuracy_op CUDA kernel

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/PaddlePaddle/Paddle/issues/4096

### CODE REVIEW

File: paddle/operators/accuracy_op.cu [hunk 1]
```diff
@@ -17,21 +17,25 @@ limitations under the License. */
...
+       i += blockDim.x * gridDim.x) {
+    for (int j = 0; j < D; ++j) {
+      if (Xdata[i * D + j] == labeldata[i]) {
+        ++count;
         break;
       }
     }
   }
-  *accuracy = static_cast<float>(correct) / static_cast<float>(N);
+  atomicAdd(&total, count);
```
Comment (dzhwinter) [inline paddle/operators/accuracy_op.cu]: LOL. It seems too complicated to me. It looks correct but I really don't have enough experience to review. Please refer to @hedaoyuan review it, or write it base on some high-level library [thrust](https://thrust.github.io).

**Comment by dzhwinter** [comment]
If you are interested in thrust, I will give a demo later.

File: paddle/operators/accuracy_op.cu [hunk 2]
```diff
@@ -17,21 +17,25 @@ limitations under the License. */
...
+  __syncthreads();
+  if (threadIdx.x == 0) {
```
Comment (hedaoyuan) [inline paddle/operators/accuracy_op.cu]: It seems that when num_samples is greater than 4096 there will be multiple blocks, and this result may be incorrect.

Comment (typhoonzero) [reply paddle/operators/accuracy_op.cu]: Thanks very much for the advise, I'll try to fix it.

Comment (typhoonzero) [reply paddle/operators/accuracy_op.cu]: Done.

File: paddle/operators/accuracy_op.cu [hunk 3]
```diff
@@ -12,26 +12,38 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-                                     const int* Xdata, const int* labelData,
-                                     float* accuracy) {
-  int correct = 0;
-  for (int row = 0; row < N; row++) {
-    const int label = labelData[row];
-    for (int col = 0; col < D; col++) {
-      const int pred = Xdata[row * D + col];
-      if (pred == label) {
-        ++correct;
+__global__ void AccuracyCudaKernel(const int N, const int D, const int* Xdata,
```
Comment (hedaoyuan) [inline paddle/operators/accuracy_op.cu]: Use `template <int BlockSize>` is better than use `PADDLE_CUDA_NUM_THREADS` macro.

File: paddle/operators/accuracy_op.cu [hunk 4]
```diff
@@ -12,26 +12,38 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+                                   const int* labeldata, float* accuracy) {
+  int count = 0;
+  __shared__ int total[PADDLE_CUDA_NUM_THREADS];
+
+  // support only 1 block
+  for (int i = threadIdx.x; i < (N); i += blockDim.x * gridDim.x) {
```
Comment (hedaoyuan) [inline paddle/operators/accuracy_op.cu]: `i += blockDim.x * gridDim.x` -> `i += BlockSize `
Use BlockSize better than blockDim.x (BlockSize is the value at compile time).

**Comment by dzhwinter** [comment]
I had tried to write a demo code to show how thrust use in that case. I find that it's also very painful to write it. 
```c++
#include <thrust/copy.h>
#include <thrust/tuple.h>
#include <thrust/device_vector.h>
#include <thrust/host_vector.h>
#include <thrust/iterator/zip_iterator.h>
#include <thrust/reduce.h>
#include <thrust/execution_policy.h>

#include <string.h>
#include <iostream>

void accuracy(int* data, int* label, int length, float* acc) {
  acc = .0;
  typedef thrust::device_vector<int>::iterator Iter;
  typedef thrust::tuple<Iter, Iter> IterPair;
  thrust::device_vector<int>  data_device(data, data+length);
  thrust::device_vector<int>  label_device(label, label+length);
  IterPair first = thrust::make_zip_iterator(thrust::make_tuple(data_device.begin(), label_device.begin()));
  IterPair last = thrust::make_zip_iterator(thrust::make_tuple(data_device.end(), label_device.end()));
  thrust::equal<thrust::tuple<int,int>> binary_op;
  thrust::device_vector<int>  correct(length);
  thrust::transform(first, last, correct.begin(), binary_op());

  int result = thrust::reduce(thrust::host, correct, correct+length);
  if (result != 0) acc = (float)result/correct.size();
}
```
related document  [zip_iterator](http://docs.nvidia.com/cuda/thrust/index.html#zip-iterator)

File: paddle/operators/accuracy_op.cu (hunk 3 above)
Comment (typhoonzero) [reply paddle/operators/accuracy_op.cu]: I would prefer `PADDLE_CUDA_NUM_THREADS`, it's a (constexpr)[http://en.cppreference.com/w/cpp/language/constexpr]. 

If use `template <int BlockSize>` we have to pass `BlockSize` twice to when calling the kernel like: `SomeKernel<BlockSize><<<1, BlockSize>>>()`, and seems `PADDLE_CUDA_NUM_THREADS` can be always the same const value.

Comment (typhoonzero) [reply paddle/operators/accuracy_op.cu]: Well, I think we can open an issue to discuss and then write some CUDA development docs.

**Comment by typhoonzero** [comment]
@dzhwinter Thanks for the very useful code! Both ways are OK, either use thrust or not, I also prefer using `thrust::reduce` it's very simple to use.

File: paddle/operators/accuracy_op.cu (hunk 3 above)
Comment (hedaoyuan) [reply paddle/operators/accuracy_op.cu]: Using the `PADDLE_CUDA_NUM_THREADS` macro is also equivalent to using it twice.

**Pull request merged**